Source/BoundaryConditions/WarpX_PEC.cpp

```diff
@@ -207,7 +206,254 @@ PEC::ApplyPECtoBfield (std::array<amrex::MultiFab*, 3> Bfield, const int lev,
                                      Bz, Bz_nodal, fbndry_lo, fbndry_hi);
             }
         );
+    }
+}
 
+
+void
+PEC::ApplyPECtoRhofield (amrex::MultiFab* rho, const int lev)
+{
+    auto& warpx = WarpX::GetInstance();
+
+    amrex::Box domain_box = warpx.Geom(lev).Domain();
+    amrex::IntVect domain_lo = domain_box.smallEnd();
+    amrex::IntVect domain_hi = domain_box.bigEnd();
+
+    amrex::GpuArray<int, 3> fbndry_lo, fbndry_hi;
+    amrex::GpuArray<ParticleBoundaryType, 3> pbndry_lo, pbndry_hi;
+    for (int idim=0; idim < AMREX_SPACEDIM; ++idim) {
+        fbndry_lo[idim] = WarpX::field_boundary_lo[idim];
+        fbndry_hi[idim] = WarpX::field_boundary_hi[idim];
+        pbndry_lo[idim] = WarpX::particle_boundary_lo[idim];
+        pbndry_hi[idim] = WarpX::particle_boundary_hi[idim];
     }
+    amrex::IntVect rho_nodal = rho->ixType().toIntVect();
+    amrex::IntVect ng_fieldgather = rho->nGrowVect();
+
+    const int nComp = rho->nComp();
+
+    amrex::Box tilebox;
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    for (amrex::MFIter mfi(*rho, amrex::TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+
+        // Extract field data
+        amrex::Array4<amrex::Real> const& rho_array = rho->array(mfi);
+
+        // Construct a tilebox to loop over the grid
+        tilebox = mfi.tilebox();
+        amrex::Box tb = convert( tilebox, rho_nodal );
```

Comment:
Not sure, but you might be able to do this in one line without the temporary `tilebox`:
```cpp
amrex::Box tb = convert(mfi.tilebox(), rho_nodal);
```